evm/src/cpu/kernel/asm/bignum/modexp.asm

```diff
@@ -0,0 +1,164 @@
+// Arithmetic on little-endian integers represented with 128-bit limbs.
+// All integers must be under a given length bound, and are padded with leading zeroes.
+
+// Stores b ^ e % m in output_loc, leaving b, e, and m unchanged.
+// b, e, and m must have the same length.
```

Comment:
It makes sense for `b` and `m` to have the same length, but I really think `e` should be separate. If `b` and `m` have length _N_ and `e` has length _M_, then this algorithm is _O_(_N_^2 _M_). However, if we were to set _N_ = _M_, then it'd be _O_(_N_^3), which is significantly worse when _N_ ≫ _M_.

As an example, say I have a really big number and I just want to square it. If we know the length of `e` then we can exit the loop way earlier.

Reply:
I guess it's fine since you call `iszero_bignum`. But it's a linear-time check, which feels a bit iffy. Either way, we can do these kinds of optimizations later.